Add documentation for overload names

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23844 Add documentation for overload names**

-

Differential Revision: [D16660167](https://our.internmc.facebook.com/intern/diff/D16660167/)

smessmer committed:
Add documentation for overload names

-

Differential Revision: [D16660167](https://our.internmc.facebook.com/intern/diff/D16660167/)

smessmer committed:
Update on "Add documentation for overload names"


-

Differential Revision: [D16660167](https://our.internmc.facebook.com/intern/diff/D16660167/)

dzhulgakov commented on aten/src/ATen/native/README.md:
```diff
@@ -170,6 +170,10 @@ Note that argument type modifiers such as defaults and optional are not currentl
 
 The declarations also support the following attributes:
 
+**Overloads.** You can register multiple functions with the same name and different
```
it renders weirdly (there's a preview button on the top right) as it's in the middle of the list of "attributes"

gchanan replied:
this is an improvement, but it still leaves some unanswered questions.  For example: do I need to provide an overload if my function has the same name?  Is there a standard naming rule for the overload?

smessmer replied:
we don't have a fixed naming scheme defined yet, apart from "try to find a good, semantically meaningful name". Is your other question whether overload names need to be present and unique if the same function name appears multiple times? The text says "you can register multiple functions [...] *if* you give them unique overload names". So the answer would be yes. Should I change that to "only if"?

gchanan replied:
I think I'd be looking for more guidance.  It's not clear to be what I _should_ do, even if it says what is possible to do.

smessmer replied:
I added some more text. Is this better?

smessmer replied:
@dzhulgakov not sure what you mean with the rendering issue. Just checked and at least now (after making the text longer), I don't find what you mean. Is it fixed or still there?

dzhulgakov replied:
It's not rendering per-se, just weird that "The declarations also support the following attributes:" starts a list of attributes (with ### header), but the first thing is not an item in the list, but just a bold text. Maybe just reorder this paragraph above the previous one?

smessmer committed:
Update on "Add documentation for overload names"


-

Differential Revision: [D16660167](https://our.internmc.facebook.com/intern/diff/D16660167/)

gchanan commented on aten/src/ATen/native/README.md:
```diff
@@ -170,6 +170,27 @@ Note that argument type modifiers such as defaults and optional are not currentl
 The declarations also support the following attributes:
 
+**Overloads.** You can register multiple functions with the same name and different
+function signatures if you give them unique overload names. An overload name
```
what happens if you try to register the same function name without overloads?

gchanan commented on aten/src/ATen/native/README.md:
```diff
@@ -170,6 +170,27 @@ Note that argument type modifiers such as defaults and optional are not currentl
+
+Overload names do not have to be globally unique, but must be unique in the set
+of all overloads for the same function. Please try to make overload names
+semantically meaningful. An overload name that just enumerates the argument
```
You should expound on what makes a name semantically meaningful.  Just saying something like "the current standard is to name the first new parameter name or type" would be helpful.

gchanan commented on aten/src/ATen/native/README.md:
```diff
@@ -170,6 +170,27 @@ Note that argument type modifiers such as defaults and optional are not currentl
+Overload names do not have to be globally unique, but must be unique in the set
+of all overloads for the same function. Please try to make overload names
+semantically meaningful. An overload name that just enumerates the argument
+types isn't helpful. Actually it is hurtful, because overload names cannot be
```
I don't get it.  The above hints that I need to give an overload name ("You can register multiple functions with the same name and different function signatures if you give them unique overload names") -- okay technically it doesn't say if and only if...

But then here it tells me if I give it a bad name it's hurtful?  But how could it be worse than something that doesn't work?

gchanan commented on aten/src/ATen/native/README.md:
```diff
@@ -170,6 +170,27 @@ Note that argument type modifiers such as defaults and optional are not currentl
+
+If you add a new function with multiple overloads, give them unique overload
+names, at most one overload is allowed to have no overload name and be the
+base variant.
```
base variant is not defined.

gchanan commented on aten/src/ATen/native/README.md:
```diff
@@ -170,6 +170,27 @@ Note that argument type modifiers such as defaults and optional are not currentl
+semantically meaningful. An overload name that just enumerates the argument
+types isn't helpful. Actually it is hurtful, because overload names cannot be
+changed for backwards compatibility reasons, even if for some reason function
+argument types change.
```
wait, I can change the argument types but I can't change the overload name?  If that's the case this needs to be explained, i.e. what is allowed to change while still being the same overload?  (compatible types from python and new default parameters?)

smessmer committed:
Update on "Add documentation for overload names"


-

Differential Revision: [D16660167](https://our.internmc.facebook.com/intern/diff/D16660167/)

[ghstack-poisoned]
